### PR TITLE
Teamleader Base Placer 🔨

### PR DESCRIPTION
**Pull request by killerswin2**
## What type of PR is this.
1. [ ] Bug
2. [ ] Change
3. [x] Enhancement

### What have you changed and why?
Information:
    Teamleader placer is a RTS style placer. The placer allows for more refined and civilized object placement.

### Please specify which Issue this PR Resolves.
Finally players can build bases.

### Please verify the following and ensure all checks are completed.

1. [x] Have you loaded the mission in LAN host?
2. [ ] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?
1. [x] No
2. [ ] Yes (Please provide further detail below.)

### How can the changes be tested?
Steps: 

********************************************************
Notes:

### CODE REVIEW

**Comment by Bob-Murphy** [comment]
@killerswin2 could you please add screenshots? That'd be nice.
Cheers

**Comment by killerswin2** [comment]
image

**Comment by killerswin2** [comment]
https://youtu.be/R9kkguX32XY

**Comment by CalebSerafin** [comment]
Yeah, this is pretty cool. I back the idea and the look of it. Has this been discussed in a meeting before, what are other people's thoughts?

**Comment by killerswin2** [comment]
We have had informal meetings about it, but we haven't sat down and discussed this in a meeting yet.

File: A3A/addons/core/functions/UtilityItems/fn_placeBuilderObjects.sqf [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+		private _price = _target getVariable ["price", 10];
+
+		private _vehicle = createVehicle [_className, [0,0,0], [], 0, "CAN_COLLIDE"];
+		_vehicle setPos _position;
+		_vehicle setDir _direction;
+		_vehicle setVariable ["price", _price, true];
+		deleteVehicle _target;
+
+		[_vehicle] call A3A_fnc_initRemoveStructure;
+		if (_className isEqualTo "Flag_AAF_F") then {
```
Comment (CalebSerafin) [inline A3A/addons/core/functions/UtilityItems/fn_placeBuilderObjects.sqf:82]: This feels rather hard-coded. Is it always valid or is there a way to detect variations of flag.

File: A3A/addons/core/functions/UtilityItems/fn_placeBuilderObjects.sqf [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        			true;
+    			} else {
+            		[_price] call A3A_fnc_resourcesPlayer;
+            		true;
+        		};
+    		};
+		};
+
+		A3A_notBuiltObjectList deleteAt (A3A_notBuiltObjectList find [_target, _objectTimeout]);
+		publicVariable "A3A_notBuiltObjectList";
```
Comment (CalebSerafin) [inline A3A/addons/core/functions/UtilityItems/fn_placeBuilderObjects.sqf:120]: We could limit the amount of "planned" constructions to minimise the potential of updating huge lists. Maybe 50 objects?

File: A3A/addons/core/functions/UtilityItems/fn_teamLeaderRTSPlacer.sqf [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+        		true;
+    		} else {
+        		if ((player getVariable ["moneyX", 0]) >= _price) then {
+            		[-_price] call A3A_fnc_resourcesPlayer;
+            		true;
+        		};
+    		};
+		};
+
+		if (_insufficientFunds) exitwith {[localize "STR_antistasi_teamleader_placer_title", localize "STR_antistasi_teamleader_placer_insufficient_funds"] call A3A_fnc_customHint};
```
Comment (CalebSerafin) [inline A3A/addons/core/functions/UtilityItems/fn_teamLeaderRTSPlacer.sqf:75]: This is executing on the client, only the server should consider insufficient funds. You'll need to fight with an SQF callback to update the UI on the client (rip other languages async await compiler sugar).

File: A3A/addons/core/functions/UtilityItems/fn_teamLeaderRTSPlacer.sqf [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+		A3A_building_EHDB set [UNSAFE_MODE, true];
+	};
+
+	if (_key isEqualTo DIK_E) then {
+		A3A_building_EHDB set [ROTATION_MODE_E, true];
+	};
+
+	if (_key isEqualTo DIK_R) then {
+			A3A_building_EHDB set [ROTATION_MODE_R, true];
+	};
```
Comment (CalebSerafin) [inline A3A/addons/core/functions/UtilityItems/fn_teamLeaderRTSPlacer.sqf:108]: User keybind configuration via CBA options soon™?

**Review by CalebSerafin — CHANGES_REQUESTED** [review]
Is did a quick skim, so I might have missed some things. But there are some comments.

File: A3A/addons/core/functions/UtilityItems/fn_placeBuilderObjects.sqf (hunk 1 above)
Comment (killerswin2) [reply A3A/addons/core/functions/UtilityItems/fn_placeBuilderObjects.sqf:82]: currently the only flag defined is that one. I could make check that rebel flag instead

File: A3A/addons/core/functions/UtilityItems/fn_placeBuilderObjects.sqf (hunk 2 above)
Comment (killerswin2) [reply A3A/addons/core/functions/UtilityItems/fn_placeBuilderObjects.sqf:120]: Yes, there will be a limit. Either based on resource allocation or just a hard limit on placement.

File: A3A/addons/core/functions/UtilityItems/fn_teamLeaderRTSPlacer.sqf (hunk 4 above)
Comment (killerswin2) [reply A3A/addons/core/functions/UtilityItems/fn_teamLeaderRTSPlacer.sqf:108]: Yes

**Comment by killerswin2** [comment]
maybe one day

**Review by jaj22 — CHANGES_REQUESTED** [review]
Ok, need the following things before I read this again:
1. Port to unstable. Save/load and utility item stuff needs switching over.
2. Sort out the indentation in teamLeaderRTSPlacerDialog.sqf.

**Comment by jaj22** [comment]
Actually I guess I should do the port myself. It was my broken shit in the end :P

**Comment by killerswin2** [comment]
I can actually push my commits

**Comment by killerswin2** [comment]
That will save the time some stuff

**Comment by jaj22** [comment]
Feel free. I haven't started yet.

**Comment by jaj22** [comment]
So what features does the placer function here have that we couldn't or wouldn't want to jam into the garage placement function?

File: A3A/addons/core/functions/UtilityItems/fn_placeBuilderObjects.sqf [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+		private _price = _target getVariable ["price", 10];
+
+		private _vehicle = createVehicle [_className, [0,0,0], [], 0, "CAN_COLLIDE"];
+		_vehicle setPos _position;
+		_vehicle setDir _direction;
+		_vehicle setVariable ["price", _price, true];
+		deleteVehicle _target;
+
+		[_vehicle] call A3A_fnc_initRemoveStructure;
+		if (_className isEqualTo A3A_faction_reb get "flag") then {
```
Comment (jaj22) [inline A3A/addons/core/functions/UtilityItems/fn_placeBuilderObjects.sqf:82]: Notes for self: Precedence error.

File: A3A/addons/core/functions/Save/fn_loadStat.sqf [hunk 6]
```diff
@@ -323,10 +323,14 @@ if (_varName in _specialVarLoads) then {
             };
             [_veh, teamPlayer] call A3A_fnc_AIVEHinit;                  // Calls initObject instead if it's a buyable item
             // TODO: Check whether various buyable items turn up as "Building"
-            if ((_veh isKindOf "StaticWeapon") or (_veh isKindOf "Building") and !isNil {_veh getVariable "A3A_canGarage"}) then {
+            if ((_veh isKindOf "StaticWeapon") or (_veh isKindOf "Building") and !isNil {_veh getVariable "A3A_canGarage"}) then { 
+                if ((!isNil "_state") && (_state isEqualType 0)) then {
+                    veh setVariable ["price", _state, true];
```
Comment (jaj22) [inline A3A/addons/core/functions/Save/fn_loadStat.sqf:328]: Notes for self: Missing underscore.

**Comment by jaj22** [comment]
Perf issues:
1. Every awaiting-construction object generates an EachFrame icon-draw handler on every player that renders regardless of distance. Should be converted to a single handler installed by initClient that renders any *nearby* awaiting-construction buildings. Can probably get away with a publicVariable array there because there likely won't be many awaiting-construction buildings at once.
2. Every awaiting-construction object has two very large addAction global remoteExecs. These can be easily reduced by remoteExec'ing a client-side function to add those actions instead.
3. Every constructed building is added to staticsToSave. Would be mitigated by the garrison rework, but I don't think there's any reason to access this list globally for buildings, so the straightforward solution for now is to move them into a separate server-side array.
4. Every constructed building has a global removal addAction remoteExec. This is potentially an awful lot of JIP remoteExec for a rarely used action, as the building count could easily be in the hundreds. Could instead have a client-side cursorObject checker: If you look at a deconstructible building then it adds the action, client-side.

**Comment by killerswin2** [comment]
On 1. That each frame handler should only be on engineers that need to see what to build.

**Comment by jaj22** [comment]
> On 1. the each-frame handler should only be on engineers who need to see what to build.

Slightly worried about players gaining unit traits later, but I'll roll with it for now.

**Comment by jaj22** [comment]
Ok, organisation/perf stuff done and working. Also added a semi-temporary GC and save filter to clear out buildings that aren't in rebel garrisons. Currently it also clears down to mission distance so there shouldn't be a risk of long-term object bloating.

Some interface issues that need to be cleaned up:
- No display of remaining funds.
- "Can't afford" hint doesn't show on top of placer UI.
- Placer UI cancel (c) doesn't do anything.
- Garage refund gives you the full 100 back.
- Plank objects not sized relative to final construction.
- Placer box object needs a stringtabled name.
- Need a second larger placer box object.
- A couple of placer-init warnings are not stringtabled yet.

**Comment by jaj22** [comment]
Added server-side locking for reasonable protection against money exploits involving two players using the same builder box or refunding a box that's in use.

Also from the to-do list:
- Now displays remaining funds in box in title bar.
- Garage refund now gives the correct amount.
- Found a decent class for the larger placer box.
- Stringtabled names for placer boxes.

Remaining:
- Better sizing of plank objects.
- Improve prices (larger bunkers are too cheap).
- Hints into the stringtable.
- Look into the cancel (or remove it). 
- Sort out logistics nodes for construction boxes.

**Comment by Bob-Murphy** [comment]
Feedback & Observations:

1. if you start building something and run away, you continue building even when the round thingy ain't visible on distance
2. `Snap to Surface` ain't working and magically let's the buildings fly away
3. `E` for clockwise rotation and `R` for counter-clockwise rotation seems weird. In the other systems we have `Q` for rotating left aka counter-clockwise and `E` for right aka clockwise.

image

4. when there are 0 credits in the building placer, the box stays but is useless. is that intended?

**Comment by jaj22** [comment]
Duh, reason for not using Q is that it's what the camera uses for up.

**Comment by jaj22** [comment]
Added functionality to cancel and repair buildings within the placer UI (so people finally have an expensive way to rebuild the towers they carelessly destroyed). Plays around with the keys a bit.

**Comment by jaj22** [comment]
Bunch more fixes/improvements done. Remaining stuff that I recall:
1. Check/expand the CUP & VN buildings.
2. Stringtables.
3. Construction box logistics nodes.

**Review by jaj22 — APPROVED** [review]

**Comment by jaj22** [comment]
Done construction box nodes and buildable objects for all maps (SPE is kinda thin, couldn't find much). Should be functionally complete now, just needs some extra stringtabling and more testing. Not sure I'll have any more time though.

**Comment by killerswin2** [comment]
I can do the string table stuff

**Comment by killerswin2** [comment]
We can do the pro-gamer move of testing on "Testing Tuesday"™️

**Comment by jaj22** [comment]
> I can do the string table stuff

Yeah feel free. I might have an hour or two available tomorrow but I'd rather spend it testing.

File: A3A/addons/core/functions/Builder/fn_buildingPlacer.sqf [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+		if (-1 != (A3A_building_EHDB # BUILD_OBJECTS_ARRAY) findIf { _x#1 == _building }) exitWith {};		// already rebuilt
+
+		// Calculate repair cost from bounding box
+		private _bbsize = (boundingBoxReal _building # 1) vectorDiff (boundingBoxReal _building # 0);
+		private _price = 6 * sqrt((_bbsize#0) * (_bbsize#1) * (_bbsize#2));
+		_price = 10 * round (_price / 10);
+
+		(_display displayCtrl IDC_PLACERHINT_T) ctrlShow true;
+		private _textCtrl = (_display displayCtrl IDC_PLACERHINT_C_TEXT);
+		_textCtrl ctrlSetText format ["T: Repair for %1 €", _price];
```
Comment (Bob-Murphy) [inline A3A/addons/core/functions/Builder/fn_buildingPlacer.sqf:219]: stringtable?

File: A3A/addons/core/functions/Builder/fn_buildingPlacer.sqf [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+		_textCtrl ctrlShow true;
+	};
+
+	if (_intersectObj in (A3A_building_EHDB # BUILD_OBJECT_TEMP_OBJECT_ARRAY)) then {
+		// show C key
+		(_display displayCtrl IDC_PLACERHINT_C) ctrlShow true;
+		private _textCtrl = (_display displayCtrl IDC_PLACERHINT_C_TEXT);
+		_textCtrl ctrlSetText format ["C: Cancel %1", getText (configof _intersectObj >> "displayName")];
```
Comment (Bob-Murphy) [inline A3A/addons/core/functions/Builder/fn_buildingPlacer.sqf:227]: stringtable?

File: A3A/addons/core/functions/OrgPlayers/fn_unitTraits.sqf [hunk 9]
```diff
@@ -44,11 +44,10 @@ else
...
- };
+    };
 };
 
-if (isMultiPlayer) then
-{
+_text spawn {
 	sleep 5;
-	["Unit Traits", format ["You have selected %1.",_text]] call A3A_fnc_customHint;
+    [localize "STR_A3A_Unit_Traits_title", format [localize "STR_A3A_Unit_Traits_selected_trait", _this]] call A3A_fnc_customHint;
```
Comment (Bob-Murphy) [inline A3A/addons/core/functions/OrgPlayers/fn_unitTraits.sqf:52]: Please revert, covered in #2970

File: A3A/addons/core/functions/UtilityItems/fn_initObjectRemote.sqf [hunk 10]
```diff
@@ -31,17 +31,41 @@ if (isNil "initClientDone") then {
 
 private _flags = (A3A_utilityItemHM get typeof _object) # 4;
 
-// movable object
-// TODO: Do we really want rotate on everything?
 if ("move" in _flags) then {
-    [_object] call A3A_fnc_initMovableObject;
+    _object addAction [
+        "Carry object",
```
Comment (Bob-Murphy) [inline A3A/addons/core/functions/UtilityItems/fn_initObjectRemote.sqf:36]: I'll deal with stringtable based on #2970

File: A3A/addons/core/functions/UtilityItems/fn_initObjectRemote.sqf [hunk 11]
```diff
@@ -31,17 +31,41 @@ if (isNil "initClientDone") then {
...
+        { [_this#3, true] call A3A_fnc_carryItem },
+        _object, 1.5, true, true, "",
+        "([_this] call A3A_fnc_countAttachedObjects == 0)
+            and (isNull attachedTo _originalTarget)", 8
+    ];
+};
+
+if ("rotate" in _flags) then {
+    _object addAction [
+        "Rotate object",
```
Comment (Bob-Murphy) [inline A3A/addons/core/functions/UtilityItems/fn_initObjectRemote.sqf:46]: I'll deal with stringtable based on #2970

File: A3A/addons/core/functions/UtilityItems/fn_initObjectRemote.sqf [hunk 12]
```diff
@@ -31,17 +31,41 @@ if (isNil "initClientDone") then {
...
 
 // loot crate object
 if ("loot" in _flags && lootToCrateRadius > 0) then {
     [_object] call A3A_fnc_initLootToCrate;
 };
 
+// building placer box
+if ("build" in _flags) then {
+    _object addAction [
+        "Building Placer",
```
Comment (Bob-Murphy) [inline A3A/addons/core/functions/UtilityItems/fn_initObjectRemote.sqf:62]: I'll deal with stringtable based on #2970

File: A3A/addons/core/Stringtable.xml [hunk 13]
```diff
@@ -9267,5 +9273,30 @@
...
+        <Original>You are not eligible to use the building placer.</Original>
+      </Key>
+      <Key ID="STR_A3A_builder_server_respond_failed">
+        <Original>Server failed to respond to building placer request.</Original>
+      </Key>
+      <Key ID="STR_A3A_builder_box_being_used">
+        <Original>Builder box is already being used by %1</Original>
+      </Key>
+    </Container>
+    <Container name="Unit_Traits">
```
Comment (Bob-Murphy) [inline A3A/addons/core/Stringtable.xml:9293]: Please revert Unit Triats strings as it's covered in #2970

**Review by Bob-Murphy — CHANGES_REQUESTED** [review]

File: A3A/addons/core/functions/Builder/fn_buildingPlacer.sqf (hunk 7 above)
Comment (jaj22) [reply A3A/addons/core/functions/Builder/fn_buildingPlacer.sqf:219]: I put these in the wrong place. I'll move them into the display UI stuff where the other related strings are.

**Comment by jaj22** [comment]
Ok, that should be everything done.

**Review by Bob-Murphy — APPROVED** [review]
lgtm

**Pull request merged**